### PR TITLE
Ports baystation projectile hitscan and visual effects

### PR DESCRIPTION
**Pull request by silicons**
Adds working hitscans to projectiles. It is default to OFF and the only way to get a gun with it on right now is to badmin varedit an ammo casing, or spawn one of the default hitscan guns.
If a projectile fired is HITSCAN it will automatically try to use the baystation visual effects including a muzzle effect from the person fired, with a tracer beam leading to the object. There's sprites for quite a few things, but I havn't made sprites for any "uncommon" guns like the tesla guns, dragnets, and stuff. ( I haven't gotten impact working yet, the impact sprite keeps "bouncing off" things like walls and mobs and generally acting wierd. That's unnessacary though.)
Adds a couple adminspawn hitscan guns. Spawn the laser tag ones for some harmless fun ;)
This should work perfectly well with things that deflect/reflect energy projectiles, only if hitscan is on the deflection/reflection is now instant so don't shoot people with energy shields with hitscan tasers if you somehow get one ;)

Code/style more or less "inspired" by baystation. Uses pixel projectile trajectory calculations. Will not work on legacy projectiles.

:cl:
experiment: Hitscan capability has been added to projectiles.
rscadd: The chameleon laser gun now has a hyper-velocity mode. Try it out!
/:cl:

### CODE REVIEW

**Comment by PKPenguin321** [comment]
might be cool for ctf

**Comment by kevinz000** [comment]
There's awefully little comments for such a big feature.

**Comment by PKPenguin321** [comment]
Admin only bullshit

If it ever gets added to the real game I'll reee

**Comment by Fox-McCloud** [comment]
Hitscan projectiles are bullshit. You don't have to aim, you just click and win--I'm leery that this PR even adds support for it, to be honest.

Bay's laser projectiles are kinda cool though.

File: code/modules/projectiles/projectile.dm [hunk 1]
```diff
@@ -36,7 +36,7 @@
 	var/nodamage = 0 //Determines if the projectile will skip any damage inflictions
 	var/flag = "bullet" //Defines what armor to use when it hits things.  Must be set to bullet, laser, energy,or bomb
 	var/projectile_type = /obj/item/projectile
-	var/range = 50 //This will de-increment every step. When 0, it will delete the projectile.
+	var/range = 100 //This will de-increment every step. When 0, it will delete the projectile.
```
Comment (phil235) [inline code/modules/projectiles/projectile.dm]: undocumented change?

File: code/modules/projectiles/projectile.dm [hunk 2]
```diff
@@ -185,8 +224,9 @@
 				new_y--
 
 			step_towards(src, locate(new_x, new_y, z))
+			var/delay = 0
```
Comment (phil235) [inline code/modules/projectiles/projectile.dm]: why the change?

File: code/modules/projectiles/projectile/beams.dm [hunk 3]
```diff
@@ -53,12 +62,18 @@
...
+	tracer_type	= /obj/effect/overlay/temp/projectile/tracer/disable
+	impact_type = /obj/effect/overlay/temp/projectile/impact/disable
 
 /obj/item/projectile/beam/pulse
 	name = "pulse"
 	icon_state = "u_laser"
 	damage = 50
 	luminosity = 2
+	muzzle_type = /obj/effect/overlay/temp/projectile/muzzle/pulse
+	tracer_type	= /obj/effect/overlay/temp/projectile/tracer/pulse
```
Comment (phil235) [inline code/modules/projectiles/projectile/beams.dm]: wrong spacing

File: code/modules/projectiles/projectile/beams.dm [hunk 4]
```diff
@@ -118,6 +133,9 @@
 /obj/item/projectile/beam/lasertag/bluetag
 	icon_state = "bluelaser"
 	suit_types = list(/obj/item/clothing/suit/redtag)
+	muzzle_type = /obj/effect/overlay/temp/projectile/muzzle/blue
+	tracer_type	= /obj/effect/overlay/temp/projectile/tracer/blue
```
Comment (phil235) [inline code/modules/projectiles/projectile/beams.dm]: wrong spacing

File: code/modules/projectiles/projectile/energy.dm [hunk 5]
```diff
@@ -17,6 +17,9 @@
 	jitter = 20
 	hitsound = 'sound/weapons/taserhit.ogg'
 	range = 7
+	muzzle_type = /obj/effect/overlay/temp/projectile/muzzle/stun
+	tracer_type	= /obj/effect/overlay/temp/projectile/tracer/stun
```
Comment (phil235) [inline code/modules/projectiles/projectile/energy.dm]: wrong spacing

File: code/modules/projectiles/projectile.dm (hunk 1 above)
Comment (kevinz000) [inline code/modules/projectiles/projectile.dm]: Addressing that tomorrow, was in during testing

File: code/modules/projectiles/projectile.dm (hunk 2 above)
Comment (kevinz000) [inline code/modules/projectiles/projectile.dm]: Earlier it was declared and set in the same line.
Now I separated those lines because if it's hitscan it doesn't need a delay.

File: code/modules/projectiles/projectile/beams.dm (hunk 3 above)
Comment (kevinz000) [inline code/modules/projectiles/projectile/beams.dm]: woops il fix that. soon. when i get up to a computer.

File: code/modules/projectiles/projectile.dm (hunk 2 above)
Comment (phil235) [inline code/modules/projectiles/projectile.dm]: Except delay is calculated two lines below, even for hitscan, so I don't really see the point of splitting.

**Comment by evsey9** [comment]
Engage in the ultimate point and click adventure.

**Comment by kevinz000** [comment]
added hitscan practice laser to chamo gun :^) also, range is back to normal

**Comment by optimumtact** [comment]
I don't know what to think, if this is merged it will inevitably find it's way into the players hands
*oranges imagines the future*

someone will add an adminonly weapon in a few months time. Later this adminonly weapon will find it's way into a ruin or a cargo crate. 

The final result will be "someone" making a security PR to add hitscan weapons to security as the default spawn weapon, the compromise will eventually be one in the armoury which the HoS will proceed to take at roundstart. 

This will imbalance things and lead to nukeops and traitors also getting hitscan weapons. Finally all weapons will be made hitscan in the controversiallly merged goofweapons overhaul.

*oranges wakes up from the nightmare*

So I'm not sure we're ready to make that jump as it really does have serious balance concerns.

**Comment by PKPenguin321** [comment]
bay features arent balanced for anything except very high RP, where most weapons aren't even used

this is really sketchy

**Comment by kevinz000** [comment]
Already added adminspawn weapons. Can remove if you want them to take the effort to click varedit.

**Comment by Supermichael777** [comment]
the hos needs a hit-scan taser

**Comment by Fox-McCloud** [comment]
Paradise has the unique position of originally being a high-leaning PR server that is now medium to medium-light RP....as such we've witnessed the transition of a lot of mechanics from Bay based to TG based.

My input on hitscan weapons is this: it takes away a lot of skill---a fight goes from maneuvering+judging distance+compensating for projectile movement+correcting for enemy movement/dodging to literally "hover mouse above target and click".

In a more roleplay oriented environment, this is probably ok, since most situations are set up to be binary win/lose in nature, not to mention things like fear, pain, and uncertainty are *forced* roleplay elements for the player.

In a less than high RP environment, though, none of those things matter, and it's just utilized as an easier point-click-win mechanic.

**Comment by Supermichael777** [comment]
#make security great again

**Comment by evsey9** [comment]
Make RnD be able to make xray hitscan pulserifles at maxed out tech.
Make hitscan guns be in the Summon Guns spell weapon rotation.
Make xray hitscan pulserifles be in the Summon Guns spell weapon rotation.
Make a firing pin that turns guns into their hitscan alternative, would require alot of tech though.

**Comment by kevinz000** [comment]
^ No.

**Comment by YakumoChen** [comment]
pretty sure /vg/ had this last I checked, and it's very much a Low RP server. Why not check and see how they're doing with it? They seem to be doing just fine.

**Comment by PKPenguin321** [comment]
yeah dude vg's excellent yeah you should totally go play there instead of here :^)

**Comment by evsey9** [comment]
I will never shut up.
But i still want hitscan guns merged
I want to engage in the ultimate point and click adventure

**Comment by Bawhoppen** [comment]
I think oranges might be right in this case, but then again a framework for interesting and unique hitscan weapons could be good

**Comment by Supermichael777** [comment]
i want it
To many greys have outrun papsmear tazer
a reckoning is due

**Comment by optimumtact** [comment]
This has to come with a weapon or some unique mechanic in game for it to be worth including and maintaining the code and then that brings all it's attendant balance concerns
